src/Microsoft.Management.Deployment/PackageManager.cpp

```diff
@@ -481,30 +473,97 @@ namespace winrt::Microsoft::Management::Deployment::implementation
                     completionEventFired = true;
                     break;
 
+                    // operationEvents[2] was signaled, operation is cancelled
+                case WAIT_OBJECT_0 + 2:
+                    cancelWaitEventFired = true;
+                    break;
+
                     // Return value is invalid.
                 default:
                     THROW_LAST_ERROR();
                 }
             }
 
-            // The install command has finished, check for success/failure and how far it got.
-            terminationHR = queueItem->GetContext().GetTerminationHR();
-            executionStage = queueItem->GetContext().GetExecutionStage();
+            if (completionEventFired)
+            {
+                // The install command has finished, check for success/failure and how far it got.
+                terminationHR = queueItem->GetContext().GetTerminationHR();
+                executionStage = queueItem->GetContext().GetExecutionStage();
+            }
         }
         WINGET_CATCH_STORE(terminationHR, APPINSTALLER_CLI_ERROR_COMMAND_FAILED);
 
         // TODO - RebootRequired not yet populated, msi arguments not returned from Execute.
-        WINGET_RETURN_INSTALL_RESULT_HR(terminationHR);
+        co_return GetInstallResult(executionStage, terminationHR, correlationData, false);
     }
 
+    winrt::Windows::Foundation::IAsyncOperationWithProgress<winrt::Microsoft::Management::Deployment::InstallResult, winrt::Microsoft::Management::Deployment::InstallProgress> GetEmptyAsynchronousResultForInstallOperation(
+        HRESULT hr,
+        hstring correlationData)
+    {
+        // If a function uses co_await or co_return (i.e. if it is a co_routine), it cannot use return directly.
+        // This helper helps a function that is not a coroutine itself to return errors asynchronously.
+        co_return GetInstallResult(::Workflow::ExecutionStage::Initial, hr, correlationData, false);
+    }
+
+#define WINGET_RETURN_INSTALL_RESULT_HR_IF(hr, boolVal) { if(boolVal) { return GetEmptyAsynchronousResultForInstallOperation(hr, correlationData); }}

```

Comment:
Why make a macro that wraps a function?  It seems like this could just be:
```suggestion
#define WINGET_RETURN_INSTALL_RESULT_HR_IF(hr, boolVal) { if(boolVal) { co_return GetInstallResult(::Workflow::ExecutionStage::Initial, hr, correlationData, false); }}
```